onadata/apps/logger/management/commands/clean_duplicated_submissions.py

```diff
@@ -4,118 +4,178 @@
-        ).filter(
-            count_uuid__gt=1
-        ).distinct()
+        # if we don't purge, we don't want to see instances
+        # that have been marked as deleted. However, if we do purge
+        # we do need these instances to be in the list in order
+        # to delete them permanently
+        if not purge:
+            query = query.filter(deleted_at=None)
+
+        query = query.values_list('uuid', flat=True)\
+            .annotate(count_uuid=Count('uuid'))\
+            .filter(count_uuid__gt=1)\
+            .distinct()
 
         for uuid in query.all():
-            instances_with_same_uuid = Instance.objects.filter(uuid=uuid).\
-                values_list('id', 'xml_hash').order_by('date_created')
+
+            duplicated_query = Instance.objects.filter(uuid=uuid)
+
+            # if we don't purge, we don't want to see instances
+            # that have been marked as deleted. However, if we do purge
+            # we do need these instances to be in the list in order
+            # to delete them permanently
+            if not purge:
+                duplicated_query = duplicated_query.filter(deleted_at=None)
+
+            instances_with_same_uuid = duplicated_query.values_list('id',
+                                                                    'xml_hash')\
+                .order_by('xml_hash', 'date_created')
             xml_hash_ref = None
             instance_id_ref = None
 
             duplicated_instances_ids = []
-            for idx, instance_with_same_uuid in enumerate(instances_with_same_uuid):
+            for instance_with_same_uuid in instances_with_same_uuid:
                 instance_id = instance_with_same_uuid[0]
                 instance_xml_hash = instance_with_same_uuid[1]
-                if idx == 0:
+
+                if instance_xml_hash != xml_hash_ref:
+                    self.__clean_up(instance_id_ref,
+                                    duplicated_instances_ids,
+                                    purge)
                     xml_hash_ref = instance_xml_hash
                     instance_id_ref = instance_id
+                    duplicated_instances_ids = []
                     continue
 
-                if instance_xml_hash == xml_hash_ref:
-                    duplicated_instances_ids.append(instance_id)
-                else:
-                    self.stdout.write('Not a duplicate #{}'.format(instance_id))
+                duplicated_instances_ids.append(instance_id)
 
+            self.__clean_up(instance_id_ref,
+                            duplicated_instances_ids,
+                            purge)
+
+        if not self.__vaccuum:
+            if purge:
+                self.stdout.write('No instances have been purged.')
+            else:
+                self.stdout.write('No instances have been marked as deleted.')
+        else:
+            # Update number of submissions for each user.
+            for user_ in list(self.__users):
```

Comment:
I think it's fine to iterate over sets without coercing them to lists. I'd be curious to know if I'm wrong about that, though :smile:

Reply:
I think you already told me that and you were right.